### DISCUSSION TITLE
Node input/output types

### DISCUSSION THREAD

**Discussion by andrewbaxter** [GraphiteEditor/Graphite#2159] [Help] [answered]
First of all, I want to say the UI is very intuitive and I'm glad it provides all the information it does (esp error messages in the node editor).

I'm trying to randomly transform nodes in a path (translate, rotate, maybe stretch handles?).  There's a couple nodes that look relevant (scatter points, jitter points, offset path?) but on the first two at least I'm getting a type error.

The path + transform nodes output `Concrete(VectorData)` - OK.  My merge nodes combining the transform outputs seem to take `Concrete(Footprint)` to `Concrete(VectorData)` AFAICT, and outputs a `Concrete(GraphicGroup)` - OK.  And Stroke takes `Concrete(Footprint)` and goes to `Concrete(GraphicGroup)` - OK.

IIUC this means there's some sort of compatibility here where `VectorData` can become a `Footprint` or `GraphicGroup` and `GraphicGroup` can become a `Footprint`, and `Footprint` can become a `GraphicGroup`.

It looks like "jitter points" needs vector data, and `GraphicGroup` can _not_ become `VectorData`.  Directly connecting the "transform" output to "jitter nodes" works (but therefore only for one path, not for the group of them).

Rambling a bit above, but
1. Would it be possible to get an overview of what these types are and how they're compatible? and
2. Is there a way to merge vector data into a vector data?
3. Kind of tangential, but is there an overview of node editor hotkeys/methods/etc? Not something polished, but just a "here's some hotkeys and workflow tips that work atm" thing or rough notes?

**Comment by Keavon** [comment, marked as answer]
> 1. Would it be possible to get an overview of what these types are and how they're compatible? and

Regarding the data types, this gets a bit technical and those types you're viewing in the tooltips and error balloons unfortunately aren't designed to be user-facing, they're more of a debug tool until we can solve a few challenges that will let us make them more user-friendly. But since you stared hard enough to try and understand it 😆, I may as well take a quick stab at an explanation:
- You can ignore the `Concrete(X)` part and just read the `X` inside, which might be:
  - `VectorData` is a vector path (containing one or more subpaths) but not multiple layers of paths— this is something akin to a single SVG element/tag, and it can only have one appearance (stroke, fill, opacity, blend mode) applied to all its subpaths
  - `GraphicGroup` is a stack of layers containing other `VectorData`, `ImageFrame` (that's raster data), or other `GraphicGroup`s (recursively forming a tree)
- All nodes have the format `() -> X` or `Footprint -> X` where `X` can be `VectorData`, `GraphicGroup`, etc. Most nodes should work with either just fine (but it used to be more awkward in this respect, and some issues may linger in less-trafficked cases).
  - `() -> X` means the node produces `X`, which is some finite data like some vector shapes or layers or whatever. Importantly, it's the *finite*, *complete* data regardless of where your viewport is currently viewing within the canvas.
  - `Footprint -> X` means the node contextually produces a momentary *perspective* into the scene's content of type `X`, such as a finite portion of an infinitely tiling pattern of tessellated vector shapes, or a finite resolution render of the raster content of a Mandelbrot set fractal, which would otherwise be infinite and impossible to represent as data. The footprint changes whenever you pan or zoom in the viewport, and the nodes get re-executed to produce a different perspective of that content.

> 2. Is there a way to merge vector data into a vector data?

The issues you're encountering are because a lot of nodes haven't yet been improved to work with group data as well as vector data when it would make sense to just apply the effect for each path in the group. (That's what issue #1929 calls for.) Currently, we have to flatten each group of paths into a single path (with multiple subpaths). You'll want to use the "Flatten Vector Elements" node for that, which converts from GraphicGroup to VectorData. In cases where you want to union the paths together into a single path, you may also want to be aware of the Boolean Operation node which also converts from GraphicGroup to VectorData but it's less robust, more slow, and you should only need it if you don't want overlapping paths for example.

> 3. Kind of tangential, but is there an overview of node editor hotkeys/methods/etc? Not something polished, but just a "here's some hotkeys and workflow tips that work atm" thing or rough notes?

Unfortunately there's nothing that lists all those details in one place. But pay attention to the contextual input hints in the status bar at the bottom of the editor, which is valuable.

**Reply by andrewbaxter** [reply]
Oh awesome, that makes sense! So it's like a collection of a collection of segments (`GraphicGroup`), and I need to go to just a collection of segments (`VectorData`).

Thanks for all the details!  And yeah, saw the status bar but it gets cut off for me due to screen size... I thought there might be something I didn't see.

**Reply by Keavon** [reply]
> Thanks for all the details! And yeah, saw the status bar but it gets cut off for me due to screen size... I thought there might be something I didn't see.

You could temporarily zoom out of the web page using your browser's scaling controls to see more on screen, as a workaround for this.
